Makefile

```diff
@@ -182,67 +190,78 @@ endif
+push-multi-arch-node-manager-image: push-all-node-images push-all-windows-node-images ## Create and push a manifest list containing all the Windows and Linux images.
 	docker manifest create --amend $(NODE_MANAGER_IMAGE) $(ALL_NODE_MANAGER_IMAGES)
-	for arch in $(LINUX_ARCHS); do \
+	for arch in $(ALL_ARCH.linux); do \
 		docker manifest annotate --os linux --arch $${arch} $(NODE_MANAGER_IMAGE)  $(NODE_MANAGER_LINUX_FULL_IMAGE):$(IMAGE_TAG)-$${arch}; \
 	done
 	# For Windows images, we also need to include the "os.version" in the manifest list, so the Windows node can pull the proper image it needs.
 	# we use awk to also trim the quotes around the OS version string.
 	set -x; \
-	for osversion in $(ALL_WINDOWS_OSVERSIONS); do \
-		full_version=`docker manifest inspect ${BASE.windows}:$${osversion} | grep "os.version" | head -n 1 | awk -F\" '{print $$4}'` || true; \
-		docker manifest annotate --os windows --arch amd64 --os-version $${full_version} $(NODE_MANAGER_IMAGE) $(NODE_MANAGER_WINDOWS_IMAGE)-$${osversion}; \
+	for windowsarch in $(ALL_ARCH.windows); do \
+		for osversion in $(ALL_OSVERSIONS.windows); do \
+			full_version=`docker manifest inspect ${BASE.windows}:$${osversion} | jq -r '.manifests[0].platform["os.version"]'`; \
+			docker manifest annotate --os windows --arch $${{windowsarch}} --os-version $${full_version} $(NODE_MANAGER_IMAGE) $(NODE_MANAGER_WINDOWS_IMAGE)-$${osversion}-$${{windowsarch}}; \
+		done; \
 	done
 	docker manifest push --purge $(NODE_MANAGER_IMAGE)
 
-# TODO(mainred): Currently we push only Linux multi-arch docker images for node image, after fully support Windows docker image building,
-#                we need to replace push-all-node-images with push-all-node-images to push multi-arch and multi-os node image, 
-#                which is tracked https://github.com/kubernetes-sigs/cloud-provider-azure/issues/829
-.PHONY: push-all-node-images
-push-all-node-images: $(addprefix push-node-image-,$(LINUX_ARCHS))
-	docker manifest create --amend $(NODE_MANAGER_IMAGE) $(ALL_LINUX_NODE_MANAGER_IMAGES)
-	for arch in $(LINUX_ARCHS); do \
-		docker manifest annotate --os linux --arch $${arch} $(NODE_MANAGER_IMAGE)  $(NODE_MANAGER_LINUX_FULL_IMAGE):$(IMAGE_TAG)-$${arch}; \
-	done
-	docker manifest push --purge $(NODE_MANAGER_IMAGE)
+.PHONY: push-all-node-images ## Push node-manager image for os and archs.
+push-all-node-images: push-all-node-images-linux push-all-node-images-windows
+
+.PHONY: push-all-node-images-linux ## Push node-manager image for Linux.
+push-all-node-images-linux: $(addprefix push-node-image-linux-,$(ALL_ARCH.linux))
+
+.PHONY: push-all-node-images-windows ## Push node-manager image for Windows.
+push-all-node-images-windows: $(addprefix push-node-image-windows-,$(ALL_OS_ARCH.windows))
+
+# split words on hyphen, access by 1-index
+word-hyphen = $(word $2,$(subst -, ,$1))
+
+push-node-image-linux-%:
+	$(MAKE) ARCH=$* push-node-image-linux
```

Comment:
Maybe we could remove such alias commands?

Reply:
The makefile target is really messy previous, and I tried to remove as many of them as possible and leave the ones exposed to customers more clear.

We want to make the Linux node images publishing easier from the customer side by calling, they don't have to do an iteration
```
push-all-node-images-linux: $(addprefix push-node-image-linux-,$(ALL_ARCH.linux))
````